pkg/lifecycle/resource_debug_enabled.go

```diff
@@ -0,0 +1,115 @@
+// +build debug_ref
+
+package lifecycle
+
+import (
+	"fmt"
+	"io"
+	"os"
+	"os/signal"
+	"path/filepath"
+	"runtime"
+	"sync"
+	"syscall"
+)
+
+// When in debug mode, we associate each reference an id and with that id the
+// stack trace that created it. We can't directly refer to the reference here
+// because we also associate a finalizer to print to stderr if a reference is
+// leaked, including where it came from if possible.
+
+func resourceClosed() error {
+	var buf [4096]byte
+	return fmt.Errorf("resource closed:\n%s", buf[:runtime.Stack(buf[:], false)])
+}
+
+// liveReferences keeps track of the stack traces of all of the live references.
+type liveReferences struct {
+	mu   sync.Mutex
+	id   uint64
+	live map[uint64][]uintptr
+}
+
+var live = &liveReferences{
+	live: make(map[uint64][]uintptr),
+}
+
+// finishId informs the liveReferences that the id is no longer in use.
+func (l *liveReferences) untrack(r *Reference) {
+	l.mu.Lock()
+	if _, ok := l.live[r.id]; ok {
+		delete(l.live, r.id)
+		runtime.SetFinalizer(r, nil)
+	}
+	l.mu.Unlock()
+}
+
+// withFinalizer associates a finalizer with the Reference that will cause it
+// to print a leak message if it is not closed before it is garbage collected.
+func (l *liveReferences) track(r *Reference) *Reference {
+	var buf [32]uintptr
+	pcs := append([]uintptr(nil), buf[:runtime.Callers(3, buf[:])]...)
+
+	l.mu.Lock()
+	r.id, l.id = l.id, l.id+1
+	l.live[r.id] = pcs
```

Comment:
Again, worth checking if the map entry already exists?

Reply:
That should be impossible due to the mutex and the increment right above it.